src/EventSubscriber/SmsMessageProcessor.php

```diff
@@ -162,9 +162,13 @@ protected function getGatewayForPhoneNumber($recipient) {
   public function deliveryReportUrl(SmsMessageEvent $event) {
     foreach ($event->getMessages() as &$sms_message) {
       if (!$sms_message->getOption('delivery_report_url')) {
-        $route_parameters = ['sms_gateway' => $sms_message->getGateway()->id()];
-        $url = Url::fromRoute('sms.process_delivery_report', $route_parameters);
-        $sms_message->setOption('delivery_report_url', $url->setAbsolute()->toString());
+        $url = $sms_message->getGateway()->getPushReportUrl();
+        try {
+          $url = $url->setAbsolute()->toString();
```

Comment:
I ended the line with `getPushReportUrl()` in case we want to optionally return NULL instead of always returning a Url object.

The route from `getPushReportUrl()` may or may not exist in the router.

The URL is actually generated, and potentially throws an exception, by `->toString()`

But there are two distinct things happening here.